synapse/storage/util/id_generators.py

```diff
@@ -316,6 +375,21 @@ def get_next_txn(self, txn: LoggingTransaction):
         txn.call_after(self._mark_id_as_finished, next_id)
         txn.call_on_exception(self._mark_id_as_finished, next_id)
 
+        # Update the `stream_positions` table with newly updated stream
+        # ID (unless self._writers is not set in which case we don't
+        # bother, as nothing will read it).
+        #
+        # We only do this on the success path so that the persisted current
+        # position points to a persited row with the correct instance name.
+        if self._writers:
+            txn.call_after(
+                run_as_background_process,
+                "MultiWriterIdGenerator._update_table",
+                self._db.runInteraction,
+                "MultiWriterIdGenerator._update_table",
+                self._update_stream_positions_table_txn,
+            )
```

Comment:
We want to do it after we've marked the associated ID as "persisted", which happens after the transaction has already finished.

Though a follow up question of: can we mark the ID as persisted before we finish the transaction? The answer to that is probably yes if we faff around a bit.

One thing to note is that the only stream that actually uses `get_next_txn` is the cache invalidation stream, which we explicitly *dont'* record the stream positions for. The only reason I've added updating the stream positions table here is for consistency and to avoid future foot guns.